engine/audiomixer.cpp

```diff
@@ -0,0 +1,167 @@
+#include "logger.h"
+#include "pong.h"
+#include "sound.h"
+
+#include <glm/glm.hpp>
+
+#include <algorithm>
+#include <cmath>
+#include <chrono>
+#include <iostream>
+
+namespace pong
+{
+
+static constexpr int SAMPLE_RATE = 44100;
+static constexpr int FRAMES_PER_BUFFER = 256;
+static constexpr glm::vec3 LEFT_LISTENER_POSITION = glm::vec3(-1125.0f, 0.0f, 0.0f);
+static constexpr glm::vec3 RIGHT_LISTENER_POSITION = glm::vec3(1125.0f, 0.0f, 0.0f);
+static constexpr float MIN_DISTANCE = 250.0f;
+
+int AudioCallbackWrapper(const void* inputBuffer, void* outputBuffer,
+                  unsigned long framesPerBuffer,
+                  const PaStreamCallbackTimeInfo* timeInfo,
+                  PaStreamCallbackFlags statusFlags,
+                  void* userData)
+{
+    return Pong::GetInstance().GetAudioMixer().AudioCallback(inputBuffer, outputBuffer, framesPerBuffer, timeInfo, statusFlags, userData);
+}
+
+bool AudioMixer::Init()
+{
+    mVolume = Config::GetValue("volume", 0.5f);
+    LogInfo("Volume: {}", mVolume);
+
+    mSpatialAudioEnabled = Config::GetValue("spatial_audio", false);
+    LogInfo("Spatial Audio: {}", mSpatialAudioEnabled ? "Enabled" : "Disabled");
+
+    PaError err = Pa_Initialize();
+    if (err != paNoError) {
+        LogError("PortAudio initialization failed: {}", Pa_GetErrorText(err));
+    }
+
+    PaStreamParameters outputParameters;
+    outputParameters.device = Pa_GetDefaultOutputDevice();
+    outputParameters.channelCount = 2; // Stereo
+    outputParameters.sampleFormat = paFloat32;
+    outputParameters.suggestedLatency = Pa_GetDeviceInfo(outputParameters.device)->defaultLowOutputLatency;
+    outputParameters.hostApiSpecificStreamInfo = nullptr;
+
+    err = Pa_OpenStream(&mStream, nullptr, &outputParameters, SAMPLE_RATE, FRAMES_PER_BUFFER, paClipOff, AudioCallbackWrapper, nullptr);
+    if (err != paNoError) {
+        LogError("PortAudio stream opening failed: {}", Pa_GetErrorText(err));
+        Pa_Terminate();
+        return false;
+    }
+
+    err = Pa_StartStream(mStream);
+    if (err != paNoError) {
+        LogError("PortAudio stream starting failed: {}", Pa_GetErrorText(err));
+        Pa_CloseStream(mStream);
+        Pa_Terminate();
+        return false;
+    }
+
+    return true;
+}
+
+void AudioMixer::Cleanup()
+{
+    PaError err = Pa_StopStream(mStream);
+    if (err != paNoError) {
+        LogError("PortAudio stream stopping failed: {}", Pa_GetErrorText(err));
+        return;
+    }
+
+    err = Pa_CloseStream(mStream);
+    if (err != paNoError) {
+        LogError("PortAudio stream closing failed: {}", Pa_GetErrorText(err));
+        return;
+    }
+
+    Pa_Terminate();
+}
+
+void AudioMixer::PlaySound(const Sound& sound)
+{
+    mPlayingSounds.emplace_back(sound);
+}
+
+void AudioMixer::PlaySound(const Sound& sound, const glm::vec3& position)
+{
+    mPlayingSounds.emplace_back(sound, position);
+}
+
+int AudioMixer::AudioCallback(const void* /*inputBuffer*/, void* outputBuffer,
```

Comment:
So a couple questions about the audio callback loop. I measured the time between callbacks and it follows a strange pattern. I would expect it to be around every 20ms right? But looking at it the time between callbacks is always something like

`~20000us, ~300us, ~300us, 20000us, 300us, 300us, and so on...`

Here is a stackoverflow question on the same thing. It gives some insight onto why, but I wasn't able to get a definitive answer. https://stackoverflow.com/questions/23214614/time-between-callback-calls

I guess I am mostly curious if you have seen something like this before? And more so, how does this effect question 2?

Reply:
And if I remeber correctly at spatial, we had the audio processing done on a timer using the event queue. So the scene enqueue a render scene timer for 20ms from now and then it would mix all the sounds in the scene. I guess I am confused as to how that was working with the pa callback? Was the buffer just assumed to be ready by the scene render timer, and then simply just put into the pa callback buffer? Does it make sense to do something like that here? 

 And if I do use something like this, does the audio processing time not only have to take <20ms every time, but also <300us every 2nd and 3rd time? Let me know if it is easier to explain this over zoom or something that might be easier.

Reply:
Question 1:

That seems reasonable enough. As an audio callback user, all you concern yourself with is providing the right number of samples back. Everything internal to PA you have to assume they are doing their best to give you the best performance and latency they possibly can. In this case, it may mean wacky "inconsistent" callbacks. Who knows why, but you have to trust the abstraction.

You may be able to mess around with this by tweaking this value, but not sure:

```
outputParameters.suggestedLatency = Pa_GetDeviceInfo(outputParameters.device)->defaultLowOutputLatency;
```

Question 2:
Your callback should take bascially no time at all, if you've done it correctly. If you get a very low time, and it is realtime safe, and you're still having issues, you should increase your latency as described above.

If you want a history of what was going on between the spatializer and the PA callback at spatial, more than happy to talk it over on zoom. It's a little complex to type (and ugly)

Reply:
Okay cool. In that case I will probably just stick with what I got.